File: test/arguments/with_range_check_code/java/with_range_check_code/VarInt32RangeCheckTest.java
package with_range_check_code;

import static org.junit.Assert.*;

import with_range_check_code.varint32_range_check.VarInt32RangeCheckCompound;

import java.io.IOException;

import org.junit.Test;

import zserio.runtime.ZserioError;
import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;

public class VarInt32RangeCheckTest
{
    @Test
    public void varInt32LowerBound() throws IOException, ZserioError
    {
        checkVarInt32Value(VARINT32_LOWER_BOUND);
    }

    @Test
    public void varInt32UpperBound() throws IOException, ZserioError
    {
        checkVarInt32Value(VARINT32_UPPER_BOUND);
    }

    @Test(expected=ZserioError.class)
    public void varInt32BelowLowerBound() throws IOException, ZserioError
    {
        checkVarInt32Value(VARINT32_LOWER_BOUND - 1);
    }

    @Test(expected=ZserioError.class)
    public void varInt32AboveUpperBound() throws IOException, ZserioError
    {
        checkVarInt32Value(VARINT32_UPPER_BOUND + 1);
    }

    private void checkVarInt32Value(int value) throws IOException, ZserioError
    {
        VarInt32RangeCheckCompound varInt32RangeCheckCompound = new VarInt32RangeCheckCompound();
        varInt32RangeCheckCompound.setValue(value);
        final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        varInt32RangeCheckCompound.write(writer);
        final byte[] writtenByteArray = writer.toByteArray();
        writer.close();

        final ByteArrayBitStreamReader reader = new ByteArrayBitStreamReader(writtenByteArray);
        final VarInt32RangeCheckCompound readVarInt32RangeCheckCompound =
                new VarInt32RangeCheckCompound(reader);
        assertEquals(varInt32RangeCheckCompound, readVarInt32RangeCheckCompound);
    }

    private static final int VARINT32_LOWER_BOUND = -268435455;
    private static final int VARINT32_UPPER_BOUND = -VARINT32_LOWER_BOUND;
}

File: test/extensions/arguments/with_range_check_code/cpp/VarInt32RangeCheckTest.cpp
#include "gtest/gtest.h"
#include "with_range_check_code/varint32_range_check/VarInt32RangeCheckCompound.h"
#include "zserio/CppRuntimeException.h"

namespace with_range_check_code
{
namespace varint32_range_check
{

class VarInt32RangeCheckTest : public ::testing::Test
{
protected:
    void checkVarInt32Value(int32_t value)
    {
        VarInt32RangeCheckCompound varInt32RangeCheckCompound;
        varInt32RangeCheckCompound.setValue(value);
        zserio::BitStreamWriter writer(bitBuffer);
        varInt32RangeCheckCompound.write(writer);

        zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
        const VarInt32RangeCheckCompound readVarInt32RangeCheckCompound(reader);
        ASSERT_EQ(varInt32RangeCheckCompound, readVarInt32RangeCheckCompound);
    }

    static const int32_t VARINT32_LOWER_BOUND;
    static const int32_t VARINT32_UPPER_BOUND;

    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);
};

const int32_t VarInt32RangeCheckTest::VARINT32_LOWER_BOUND = -static_cast<int32_t>((1U << 28U) - 1);
const int32_t VarInt32RangeCheckTest::VARINT32_UPPER_BOUND = static_cast<int32_t>((1U << 28U) - 1);

TEST_F(VarInt32RangeCheckTest, varInt32LowerBound)
{
    checkVarInt32Value(VARINT32_LOWER_BOUND);
}

TEST_F(VarInt32RangeCheckTest, varInt32UpperBound)
{
    checkVarInt32Value(VARINT32_UPPER_BOUND);
}

TEST_F(VarInt32RangeCheckTest, varInt32BelowLowerBound)
{
    try
    {
        checkVarInt32Value(VARINT32_LOWER_BOUND - 1);
        FAIL() << "Actual: no exception, Expected: zserio::CppRuntimeException";
    }
    catch (const zserio::CppRuntimeException& excpt)
    {
        ASSERT_STREQ("Value -268435456 of VarInt32RangeCheckCompound.value exceeds the range of "
                     "<-268435455..268435455>!",
                excpt.what());
    }
}

TEST_F(VarInt32RangeCheckTest, varInt32AboveUpperBound)
{
    try
    {
        checkVarInt32Value(VARINT32_UPPER_BOUND + 1);
        FAIL() << "Actual: no exception, Expected: zserio::CppRuntimeException";
    }
    catch (const zserio::CppRuntimeException& excpt)
    {
        ASSERT_STREQ("Value 268435456 of VarInt32RangeCheckCompound.value exceeds the range of "
                     "<-268435455..268435455>!",
                excpt.what());
    }
}

} // namespace varint32_range_check
} // namespace with_range_check_code
